### ISSUE TITLE
Generic method argument with Self type causes incomplete signature

### ISSUE THREAD

**Issue by emmatyping** [microsoft/pyright#10636]
**Describe the bug**
pyright seems to mis-infer the signature of a method when an parameter to a generic argument is a self-type that is checked against a protocol.

**Code or Screenshots**
```python
from __future__ import annotations

from typing import Any, Protocol, Self

class _ViewP(Protocol):
    # XXX: Any should really be some sort of `Pipeline`-bound TypeVar...
    def dispatch(self, pipeline: Any) -> None:
        ...

class PipelineView:
    def dispatch(self, pipeline: Pipeline[Self]) -> None:
        ...

class Pipeline[ViewT: _ViewP]:
    def pipeline_views(self) -> ViewT:
        ...
```

This produces
```
repro.py:11:43 - error: Type "Self@PipelineView" cannot be assigned to type variable "ViewT@Pipeline"
    Type "PipelineView*" is not assignable to upper bound "_ViewP" for type variable "ViewT@Pipeline"
      "PipelineView*" is incompatible with protocol "_ViewP"
        "dispatch" is an incompatible type
          Type "() -> Unknown" is not assignable to type "(pipeline: Any) -> None"
            Function accepts too many positional parameters; expected 0 but received 1 (reportInvalidTypeArguments)
1 error, 0 warnings, 0 informations
```

**VS Code extension or command-line**
Using `pyright 1.1.402` through the CLI.
